File: Design-Patterns/Visitor/CSharp/Visitors/DropEquipmentVisitor.cs
using System;
using Visitor.Equipments;

namespace Visitor.Visitors
{
    public class DropEquipmentVisitor : EquipmentVisitor
    {
        private readonly Equipment _equipment;
        private readonly Backpack _backpack;

        public DropEquipmentVisitor(Equipment equipment, Backpack backpack)
        {
            _equipment = equipment;
            _backpack = backpack;
        }

        public override void VisitEquipment(Equipment equipment)
        {
            if (equipment == _equipment)
            {
                Console.WriteLine($"Drops equipment \"{equipment.Name}\".");
                _backpack.Remove(equipment);
                GameSystem.Instantiate(equipment);
            }
        }

        public override void VisitArmor(Armor armor)
        {
            if (armor == _equipment)
            {
                Console.WriteLine($"Drops equipment \"{armor.Name}\". Creates slash effect.");
                _backpack.Remove(armor);
                GameSystem.Instantiate(armor);
            }
        }

        public override void VisitWeapon(Weapon weapon)
        {
            if (weapon == _equipment)
            {
                Console.WriteLine($"Drops equipment \"{weapon.Name}\". Creates flash effect.");
                _backpack.Remove(weapon);
                GameSystem.Instantiate(weapon);
            }
        }
    }
}

File: Design-Patterns/Visitor/Cpp/drop_equipment_visitor.cpp
#include "equipment_visitor.h"
#include <iostream>
#include "game_system.h"

drop_equipment_visitor::drop_equipment_visitor(std::shared_ptr<equipment> equipment, std::shared_ptr<backpack> backpack)
{
	equipment_ = equipment;
	backpack_ = backpack;
}

void drop_equipment_visitor::visit_equipment(std::shared_ptr<equipment> equipment)
{
	if (equipment == equipment_)
	{
		std::cout << "Drops equipment \"" + equipment->get_name() + "\"." << std::endl;
		backpack_->remove(equipment);
		game_system::instantiate(equipment.get());
	}
}

void drop_equipment_visitor::visit_weapon(std::shared_ptr<weapon> weapon)
{
	if (weapon == equipment_)
	{
		std::cout << "Drops weapon \"" + weapon->get_name() + "\". Creates slash effect." << std::endl;
		backpack_->remove(weapon);
		game_system::instantiate(weapon.get());
	}
}

void drop_equipment_visitor::visit_armor(std::shared_ptr<armor> armor)
{
	if (armor == equipment_)
	{
		std::cout << "Drops equipment \"" + armor->get_name() + "\". Creates flash effect." << std::endl;
		backpack_->remove(armor);
		game_system::instantiate(armor.get());
	}
}
